src/cpu.zig

```diff
@@ -2,210 +2,123 @@ const std = @import("std");
+    context: anytype,
+    comptime resultCallback: fn (@TypeOf(context), slimy.Result) void,
+    comptime progressCallback: ?fn (@TypeOf(context), completed: u64, total: u64) void,
+) !void {
+    std.debug.assert(params.method == .cpu);
+    std.debug.assert(params.method.cpu > 1);
+    std.debug.assert(params.x0 < params.x1);
+    std.debug.assert(params.z0 < params.z1);
+
+    // Reset chunk search counter
+    chunks_searched = std.atomic.Value(usize).init(0);
+
+    var threads = std.BoundedArray(std.Thread, 255).init(0) catch unreachable;
+    const thread_count = params.method.cpu;
+    for (0..thread_count) |thread_index| {
+        threads.append(try std.Thread.spawn(
+            .{ .stack_size = 64 * 1024 },
+            worker,
+            .{
+                params,
+                context,
+                resultCallback,
+                progressCallback,
+                thread_index,
+                thread_count,
+            },
+        )) catch unreachable;
+    }
+    std.Thread.yield() catch {};
+    for (threads.slice()) |thread| {
+        thread.join();
     }
-    return count;
 }
 
-pub fn search(
+pub fn worker(
     params: slimy.SearchParams,
     context: anytype,
     comptime resultCallback: fn (@TypeOf(context), slimy.Result) void,
     comptime progressCallback: ?fn (@TypeOf(context), completed: u64, total: u64) void,
+    thread_id: usize,
+    thread_count: usize,
 ) !void {
-    try Searcher(struct {
-        ctx: @TypeOf(context),
-
-        const Self = @This();
-
-        pub fn reportResult(self: Self, result: slimy.Result) void {
-            resultCallback(self.ctx, result);
-        }
-        pub fn reportProgress(self: Self, completed: u64, total: u64) void {
-            if (progressCallback) |callback| {
-                callback(self.ctx, completed, total);
-            }
+    const block_size = SearchBlock.tested_size;
+
+    const blocks_x = try std.math.divCeil(usize, @intCast(params.x1 - params.x0), block_size);
+    const blocks_z = try std.math.divCeil(usize, @intCast(params.z1 - params.z0), block_size);
+
+    // split blocks as evenly as possible
+    const start_block = blocks_x * blocks_z * thread_id / thread_count;
+    const end_block = blocks_x * blocks_z * (thread_id + 1) / thread_count;
+
+    var i: usize = 0;
+    for (start_block..end_block) |block_index| {
+        const rel_block_x = block_index / blocks_x;
+        const rel_block_z = @mod(block_index, blocks_x);
+        var chunk = SearchBlock.initSimd(params.world_seed, params.x0 + @as(i32, @intCast(rel_block_x * block_size)), params.z0 + @as(i32, @intCast(rel_block_z * block_size)));
+        chunk.preprocess();
+        _ = chunk.calculateSliminess(params, context, resultCallback);
+        i += 1;
+        if (i == 20) {
+            _ = chunks_searched.fetchAdd(i, .monotonic);
+            i = 0;
+            if (thread_id == 0) (progressCallback orelse continue)(context, chunks_searched.raw, blocks_x * blocks_z);
```

Comment:
this is a... creative use of `continue`

I'd probably recommend just doing this instead:

```zig
if (thread_id == 0 and progressCallback != null) {
    progressCallback.?(...);
}
```